Basic named parameters

snoe925 opened the pull request:
Add -named-parameters to the proc command using -- in the proc arguments.  The -- indicates that the proc uses named parameters.  All proc parameters before the -- are named and all following are positional.  At proc compile time, the -- flags the proc as a named parameter proc.  Therefore, at runtime non-named parameters only have 1 bit check of overhead.  At runtime the int objc, Tcl_Obj * argv are rewritten to a new list matching the position parameter invocation of the proc.

       proc name {named1 named2 -- pos1 pos2 args} body

For example,

      proc two_required_one_optional {req1 req2 {opt3 def3} -- pos1 {pos2 def4}} ...

The -- cannot be the first parameter in the proc.  That would be redundant with TCL's normal positional parameters.

snoe925 committed:
Merge pull request #1 from TCL-2020/core-8-branch

Merge upstream

Commit:
WIP: named parameters

Rewrite the TCL call frame using named parameter syntax:

    proc -named-parameters A {{a a_def} b {c c_def}} {...}
    A -a 1 -b 2 -c 3
    A -b 2 -a 1
    # will not work outside of proc
    A -b 2
    # b is required
    A -a 1 -c 2

If a parameter is missing and the proc does not have a default
value declared, then that is an error.

The script interpreter checks the pops from the call stack.
Currently this will only work inside of a proc if the parameter list
grows because of parameter expansion.

Commit:
Rewrite to pass tests

Rewrite to handle all the tests. Passes tests
but does not have proper memory handling.

Commit:
Remove stray variable decl

Commit:
Use a TCL DString for error messages

Use a TCL DString to create the error message.

Commit:
Allocate new objv using normal TclpAlloc

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,229 @@ InitArgsAndLocals(
+    }
+    markerIndex++;
+    newObjv[markerIndex] = Tcl_NewStringObj("--", 2);
+    /* For each passed argument, either copy to positional argv array. */
```
"either copy to positional argv array"... or what... what's the other half of the "either"?

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,229 @@ InitArgsAndLocals(
+                /* we have to grow the newObjv for the args parameters */
+                size_t oldSize = newObjc * sizeof(*newObjv);
+                newObjc = newObjc + (framePtr->objc - srcIndex);
+                Tcl_Obj ** newNewObjv = (Tcl_Obj **)Tcl_AttemptAlloc(newObjc * sizeof(*newObjv));
```
Tcl_AttemptAlloc can fail.  You need to either confirm that it succeeded or replace the call with Tcl_Alloc.

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,229 @@ InitArgsAndLocals(
+        srcIndex++;
+    }
+
+    /* for any parameters not provided in the argument list supply the default
```
"for any parameters not provided in the argument list that have a default, supply the default
if a named parameter or generate an error if a positional parameter is not provided and doesn't have a default"

lehenbauer reviewed (changes requested):
Looks good.  I think the Tcl_AttemptAlloc needs to be fixed.  Other remarks are minor.

Commit:
Use Tcl_ListObj* APIs for parameters.

Build the new call parametes with a Tcl List like
ensembles rewrite calls.  Then we can implement the
free of the new argv with a NRCallback.

snoe925 commented:
To get the memory handling correct I had to make a Tcl List for the new call parameters.  That is easier if the list operations are appends.  Therefore, I had to rewrite most of the logic.
But overall I think it's a bit more clear what is going on.

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,337 @@ InitArgsAndLocals(
+
+    /* Internal error. We did not find the marker. */
+    if (markerIndex == 0) {
+        return TCL_ERROR;
```
Would hitting this mean there is a bug in the named parameter software?  If so then maybe make it be an assertion instead?

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,337 @@ InitArgsAndLocals(
+        while (ii < framePtr->procPtr->numArgs && varPtr) {
+            const char * const name = varPtr->name;
+            /* check if name a parameter is -- */
+            if (markerIndex == 0 && name[0] == '-' && name[1] == '-' && !name[2]) {
```
I might've used "name[2] != '\0'" here for readability rather than "!name[2]", but no biggie.

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,337 @@ InitArgsAndLocals(
+
+    /* Find where the -- is located in the proc declaration.
+     * And determine if the proc has variable number of arguments.
+     * Caching this information when compiling the proc would be faster. */
```
So every time any proc is called this code will loop through the arguments defined in the proc to look for --?

lehenbauer commented on generic/tclProc.c:
```diff
@@ -1513,6 +1536,337 @@ InitArgsAndLocals(
+        int newObjc;
+
+        /* Get the pointer to the object array */
+        Tcl_ListObjGetElements(NULL, copyPtr, &newObjc, &newObjv);
```
Tcl_ListObjGetElements can fail and return TCL_ERROR.  Perhaps that's not possible here, though.

lehenbauer reviewed (approved):
Approved... but please read through my comments and see if there's anything you think you should change.

resuna commented:
I still think we need to be able to have procs that accept both positional and named parameters, because it's such a commonly useful idiom. (eg gets, switch, etc).

proc --named-parameters foo {positional1 positional2 {-named default} ...} { ... }

That way you can use the number of parameters to tell if there's a need for a -- and won't always need a -- unless it's actually ambiguous.
